### PR TITLE
Deal with dirty encoder or broken phase as non-HF errors

### PR DESCRIPTION
**Pull request by Nicogene**
These errors used to trigger HW, now just print out the problem

Please review code.

cc @S-Dafarra @pattacini

### CODE REVIEW

**Comment by S-Dafarra** [comment]
Thanks @Nicogene! Just to understand, the message will appear also as a warning in the logger right?

**Comment by S-Dafarra** [comment]
> It will appear an error, but no HW fault should be triggered
> 
> We can change it to warning but since the robot should be stopped at some point maybe is better to mark it as a problem

No need to change, indeed I prefer if it still appears as an error. I was trying to understand the code 😁

**Comment by pattacini** [comment]
This PR will be applied FW-wide to all robots, in my understanding.
Do we want to go this way? What would it be the impact on for example iCub 2.x?

Assessing this is not easy, I think. Probably, we could go for an experimental FW to address @S-Dafarra needs first.
What do you think @S-Dafarra @Nicogene ?

**Comment by pattacini** [comment]
After a discussion with @ale-git, we think it's safe to proceed with merging the PR also considering the iCub 2.5 robots.

**Comment by S-Dafarra** [comment]
cc @isorrentino

**Review by marcoaccame — CHANGES_REQUESTED** [review]
> After a discussion with @ale-git, we think it's safe to proceed with merging the PR also considering the iCub 2.5 robots.

ok, i agree.

however, we should also update the version of the following boards so that we can always understand which fw versions are running on the robots: `2foc`, `ems`. maybe also `mc4plus` and `mc2plus` as they use the same code in Motor.c

and also we need to produce new binaries to be placed in `icub-firmware-build`.

pls contact me if you need assistance in that.

**Comment by GiulioRomualdi** [comment]
Hi all, we tried to test the firmware. We flashed the 2foc and everything went fine and we were able to run the robot.

Then we flashed the ems and we got the following error while running the `yarprobotinterface`. 
We have the latest version of `icub-firmware-shared` and `icub-main` of devel. 
```
[1611859326.601956] [ERROR] /usr/local/src/robot/robotology-superbuild/robotology/ICUB/src/libraries/icubmod/embObjLib/ethResource.cpp:298 bool eth::EthResource::verifyBoardTransceiver() (0x0000314f) EthResource::verifyBoardTransceiver() detected different mn protocol minor versions: local = 17 , remote = 16 : cannot proceed any further.
[1611859326.602063] [ERROR] /usr/local/src/robot/robotology-superbuild/robotology/ICUB/src/libraries/icubmod/embObjLib/ethResource.cpp:299 bool eth::EthResource::verifyBoardTransceiver() (0x0000314f) ACTION REQUIRED: BOARD right_arm-eb29-j12_15 with IP 10.0.1.29 needs a FW update.
[1611859326.602191] [ERROR] /usr/local/src/robot/robotology-superbuild/robotology/ICUB/src/libraries/icubmod/embObjLib/ethResource.cpp:499 virtual bool eth::EthResource::verifyEPprotocol(eOprot_endpoint_t) (0x0000314f) EthResource::verifyEPprotocol() cannot verify BOARD right_arm-eb29-j12_15 with IP 10.0.1.29 : cannot proceed any further
[1611859326.602327] [ERROR] /usr/local/src/robot/robotology-superbuild/robotology/ICUB/src/libraries/icubmod/embObjMotionControl/embObjMotionControl.cpp:376 virtual bool yarp::dev::embObjMotionControl::open(yarp::os::Searchable&) (0x0000314f) embObjMotionControl: failed verifyEPprotocol. Cannot continue!
```

**Comment by isorrentino** [comment]
We understood which was the problem. After flashed the EMS boards, we needed to update `icub_firmware_shared` to the latest devel. After that the firmware of the mc4plus boards was unable to start and the `yarprobotinterface` was requesting a firmware update. Then, we flashed also the mc4plus boards with the latest devel and the `yarprobotinterface` started but it prints on the terminal the following error saturating immediatly the logger:

Screenshot from 2021-01-29 11-16-42
  
Now, we are going to test the walking.

**Comment by S-Dafarra** [comment]
> We understood which was the problem. After flashed the EMS boards, we needed to update `icub_firmware_shared` to the latest devel. After that the firmware of the mc4plus boards was unable to start and the `yarprobotinterface` was requesting a firmware update. Then, we flashed also the mc4plus boards with the latest devel of `icub-firmware-build` and the `yarprobotinterface` started but it prints on the terminal the following error saturating immediatly the logger:
> 
> Screenshot from 2021-01-29 11-16-42
> 
> Now, we are going to test the walking.

Interestingly, it is happening on the right leg. I wonder if it is related to https://github.com/icub-tech-iit/task-force-miscellanea/issues/69. I also wonder how the encoder can be broken and the motor still be able to move.

**Comment by isorrentino** [comment]
the Walking is working fine

**Comment by Nicogene** [comment]
> We understood which was the problem. After flashed the EMS boards, we needed to update icub_firmware_shared to the latest devel. After that the firmware of the mc4plus boards was unable to start and the yarprobotinterface was requesting a firmware update.

I suspected that might be the problem, I created the binaries using the latest `icub-firmware-shared`, good that you solved it!
About the print, the idea is to keep the error printed but not trigger HW fault. 

> the Walking is working fine

Super!
I hope that the encoder is actually broken and it is not a false error

**Comment by pattacini** [comment]
Good that the walking is working 👍🏻 

>Interestingly, it is happening on the right leg. I wonder if it is related to icub-tech-iit/task-force-miscellanea#69. I also wonder how the encoder can be broken and the motor still be able to move.

This PR just disables the HF to raise error messages instead, so I'd expect that the HF was triggered formerly.
Is that correct? Maybe not as continuously as it seems now.

**Comment by pattacini** [comment]
For this PR, @Nicogene could you please address @marcoaccame 's comments and then proceed with merging/uploading the FW?

**Comment by Nicogene** [comment]
> For this PR, @Nicogene could you please address @marcoaccame 's comments and then proceed with merging/uploading the FW?

For the ems I can figure out how to change the version, but I might need help for the 2foc, @marcoaccame I will contact you!

**Comment by isorrentino** [comment]
@Nicogene here is an example on how to change the 2foc firmware version https://github.com/isorrentino/icub-firmware/commit/45951ab6d7716b038bc7a1450c5a096d41f19e2f.

**Comment by Nicogene** [comment]
@marcoaccame I increased the version of all the board you mentioned
Did I do it correctly?

**Comment by marcoaccame** [comment]
> @marcoaccame I increased the version of all the board you mentioned
Did I do it correctly?


let me check ...

**Comment by marcoaccame** [comment]
ok, correct

**Review by marcoaccame — APPROVED** [review]
ok

**Comment by GiulioRomualdi** [comment]
Could you push the new version of the firmware on icub-firmware-build? So we can update it with the latest modifications?

**Comment by pattacini** [comment]
@Nicogene could you do that?

**Comment by Nicogene** [comment]
I can do it for ems, mc4plus and mc2plus, but I don't know how to build the 2foc firmware

**Comment by pattacini** [comment]
@marcoaccame @ale-git any clue on the 2FOC?

**Comment by isorrentino** [comment]
@Nicogene I think this comment https://github.com/dic-iit/element_torque-control-via-current/issues/126#issuecomment-658288727 can help you (I dont' know if you have the rights to read it). I copy the comment here.

> ## Installing the MPLAB X IDE
> 
> - Goto https://www.microchip.com/mplab/mplab-x-ide and browse down to "Downloads" section and download the latest version of MPLAB® X IDE driver for Windows (currently v5.40).
> - follow the instruction and default options
> 
> ## Installing the MPLAB XC Compilers
> 
> All the compilers and related dependencies can be downloaded from https://www.microchip.com/mplab/compilers, in the last section, in the tab "Compiler Downloads".
> image
> 
> The latest version for the compiler and dependencies on Windows is **v1.50**.
> 
> - After the install of the MPLAB X IDE is complete, follow the suggestion to download the "MPLAB XC compiler 16 bits" for dsPICs, and download the "MPLAB XC16 Compiler vX.XX" compiler for Windows, in "Compiler Downloads" tab (as shown above), "Compilers" subsection.
> - Launch the compiler installation: select the "free" option, and select the default options in the steps that follow.
> - Download Install the "MPLAB XC16 Compiler Part-Support Patch vX.XX" from the subsection "Latest Part Support Patch Files". This is a priori for supporting new parts, but actually the dsPIC used in our setup, version **33FJ128MC802**, appears in the "supported parts in the update" secttion of the [release notes of the patch](https://ww1.microchip.com/downloads/en/DeviceDoc/XC16-v1.50-part-support-release-notes.html).
> - Download Install the "Current Peripheral Libraries" -> "16-bit dsPIC33, PIC24E, PIC24H MCUs: Legacy Peripheral Libraries" from the subsection "Peripheral Libraries (PLIBS)" (latest version of "PIC24/dsPIC Peripheral Libraries v2.00 - Windows" [here](https://www.microchip.com/mymicrochip/filehandler.aspx?ddocname=en574967)).
> 
> ### Note on the "MPLAB XC16 Compiler vX.XX" compiler
> 
> The compiler "MPLAB XC16 Compiler v1.50" alone is missing some dependencies like the include `<pwm12.h>`. If you build the project without the mentioned additonnal patches, you'll get the error:
> ```
> CLEAN SUCCESSFUL (total time: 8ms)
> make -f nbproject/Makefile-default.mk SUBPROJECTS= .build-conf
> make[1]: Entering directory 'C:/Users/icub/git/icub-firmware/emBODY/eBcode/arch-dspic/board/2foc/appl/2FOC-V3/2FOC-V3.X'
> make  -f nbproject/Makefile-default.mk dist/default/production/2FOC-V3.X.production.hex
> make[2]: Entering directory 'C:/Users/icub/git/icub-firmware/emBODY/eBcode/arch-dspic/board/2foc/appl/2FOC-V3/2FOC-V3.X'
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/2FOC.c  -o build/default/production/_ext/1360937237/2FOC.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/2FOC.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/DHES.c  -o build/default/production/_ext/1360937237/DHES.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/DHES.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../asm/MeasCurr.s  -o build/default/production/_ext/1360919980/MeasCurr.o  -c -mcpu=33FJ128MC802  -omf=elf -DXPRJ_default=default  -legacy-libc  -I"../asm" -Wa,-MD,"build/default/production/_ext/1360919980/MeasCurr.o.d",--defsym=__MPLAB_BUILD=1,--no-relax  -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/Faults.c  -o build/default/production/_ext/1360937237/Faults.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/Faults.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/PWM.c  -o build/default/production/_ext/1360937237/PWM.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/PWM.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/ADC.c  -o build/default/production/_ext/1360937237/ADC.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/ADC.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/System.c  -o build/default/production/_ext/1360937237/System.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/System.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/DCLink.c  -o build/default/production/_ext/1360937237/DCLink.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/DCLink.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> "C:\Program Files\Microchip\xc16\v1.50\bin\xc16-gcc.exe"   ../src/can_icubProto.c  -o build/default/production/_ext/1360937237/can_icubProto.o  -c -mcpu=33FJ128MC802  -MMD -MF "build/default/production/_ext/1360937237/can_icubProto.o.d"        -omf=elf -DXPRJ_default=default  -legacy-libc    -ffunction-sections -mlarge-data -O3 -I"../include" -msmart-io=1 -Wall -msfr-warn=off    -mdfp="C:/Program Files (x86)/Microchip/MPLABX/v5.40/packs/Microchip/dsPIC33F-GP-MC_DFP/1.2.35/xc16"
> Options have been disabled due to restricted license
> Visit https://www.microchip.com/compilers to purchase a new key.
> In file included from ../src/Faults.c:5:0:
> ../include/PWM.h:10:19: fatal error: pwm12.h: No such file or directory
> compilation terminated.
> ```
> This is a known issue with the latest drivers released after version **v1.24**. The fast workaround in the previous setuhttps://github.com/robotology/icub-firmware/issues/161p was to revert the driver back to version 1.24. That version should be available in https://www.microchip.com/development-tools/pic-and-dspic-downloads-archive, but does not clearly appear in section "MPLAB C Compiler for PIC24 and dsPIC DSCs" and seems to be available only as "Source Archives". The real cause of the issue is actually explained in the [v1.25 release notes](http://ww1.microchip.com/downloads/en/DeviceDoc/XC16-v-1.25%20release%20notes.html):
> ```
> The peripheral libraries are no longer shipped with the compiler package. These may be downloaded separately from the MPLAB XC download page.
> ```
> So, the proper fix is to download the additional Legacy Peripheral Libraries as explained int the previous section.
> 
> ## Running the MPLAB X IDE [WIP]
> 
> - Run app MPLAB X IDE v5.40.
> - Open the project [2FOC-V3.X](https://github.com/robotology/icub-firmware/tree/master/emBODY/eBcode/arch-dspic/board/2foc/appl/2FOC-V3/2FOC-V3.X) (shortcut CTRL+SHIFT+O on windows).
> - you should see in the lower left pane:
> 
> image
> 
> The 2FOC device is the **dsPIC33FJ128MC802**.
> - Then click on Packs - dsPIC33F-GP-...   You should see a popup window with the project properties
> 
> image
> 
> just click "OK".
> - Clean and Build the project as shown below:
> image
> 


If you include also @GiulioRomualdi's modifications you can compile it also on Ubuntu (https://github.com/GiulioRomualdi/icub-firmware/commit/e31a04b6620d06809c7897d781bf263ad1a90b79).


I think we should open an issue similar to https://github.com/robotology/icub-firmware/issues/161.

cc @pattacini

**Comment by Nicogene** [comment]
@ale-git compiled for me the 2foc firmware, here is the PR for integrating it in `icub-firmware-build` https://github.com/robotology/icub-firmware-build/pull/20

File: emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.h [hunk 1]
```diff
@@ -207,6 +207,7 @@ typedef struct //Motor
     BOOL wrong_ctrl_mode;
     
     uint16_t diagnostics_refresh;
+    uint16_t diagnostics_refresh_warning;
```
Comment (pattacini) [inline emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.h:210]: Hi @Nicogene 

I've tried to seek around but this variable doesn't seem to be initialized anywhere.

File: emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.c [hunk 2]
```diff
@@ -670,8 +671,18 @@ BOOL Motor_check_faults(Motor* o) //
         {
             o->can_dead = FALSE;
         }
+
+        // No HW fault triggered by this warnings
+        if (++o->diagnostics_refresh_warning > 5*CTRL_LOOP_FREQUENCY_INT) {
```
Comment (pattacini) [inline emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.c:676]: @Nicogene, what's the logic of this check?

This is the only place where `o->diagnostics_refresh_warning` seems to get used.
However, it doesn't appear to be reset anywhere else as it happens for its counterpart `o->diagnostics_refresh` from which you most likely inherited the usage.

**Review by pattacini — COMMENTED** [review]
Hi @Nicogene 

I've got a couple of doubts on this, although the PR was merged several days ago.

I'm here to just pave the way to the request for muting the continuous stream of error messages (after a certain number is sent).

File: emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.c (hunk 2 above)
Comment (Nicogene) [reply emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.c:676]: You are right, I forgot to put o->diagnostics_refresh = 0; inside this if block

Comment (pattacini) [reply emBODY/eBcode/arch-arm/embobj/plus/mc/Motor.c:676]: Don't worry, we need to revise the way these messages are handled anyhow.

**Pull request merged**